### ISSUE TITLE
Batch mode over a graph

### ISSUE THREAD

**Issue by fabienbaradel** [dmlc/dgl#304]
Hi, I have some question about how to use a graph on a batch of examples. Here is an example:
```
# Create a graph and init
g = dgl.DGLGraph()
N, D = 4, 3
g.add_nodes(N)  # add 4 nodes
edge_list = [(0, 1), (1, 2), (2, 3)]
src, dst = tuple(zip(*edge_list))
g.add_edges(src, dst)  # link from src to dst
g.add_edges(dst, src)  # and bi-directional
E = len(edge_list) * 2 # because bi-directionnal

# No batch mode
g.ndata['h'] = torch.randn(N, D)
g.edata['h'] = torch.randn(E, E)

# Batch mode - share the graph over different example: correct ?
B = 16
bg = dgl.batch([g]*B)
bg.ndata['h'] = torch.randn(B*N, D)
g.edata['h'] = torch.randn(B*E, E)
```

Is it correct that by using `dgl.batch` I am sharing the graph between all examples in my graph?
Otherwise could you indicate me how to use it in order to do batch mode?

Thanks in advance

Fabien

**Comment by zhangyu2ustc** [comment]
@fabienbaradel  I am facing the same problems here. Can you please share with me how did you slove the problem? I am not show that stacked a list of graphs is right option...

**Comment by jermainewang** [comment]
@zhangyu2ustc please checkout this tutorial: https://docs.dgl.ai/tutorials/basics/4_batch.html

**Comment by zhangyu2ustc** [comment]
@jermainewang Thanks! I am not totally understand how to generalize to my case. In my projects, I was trying to predict the labels from samples which includes a feature vector on each node of the graph. Thus, the graph is fixed among all samples. Do I need to duplicate the graph structure Batch-size times? It does not seem reasonable because I have a very large graph to start with (96k nodes).  with batch-size=128, it will turns into a huge graph. But all I need to do is to pass the samples with size (batch_size, graph_num_of_node, feature_dim). Could you explain in more details how should I do it? Thanks again!
Best,
   Yu.

**Comment by BarclayII** [comment]
The tutorial talks about batch-processing multiple graphs.

Are you instead batching *nodes on the same graph*?

**Comment by zhangyu2ustc** [comment]
@BarclayII Thanks. I am not exactly batching on nodes. I am working on a supervised learning problem, where each sample with size (1, graph_num_of_node, feature_dim) has a label to be learned. I want batching on a group of samples. The graph itself is fixed but with a multi-layer architecture. 

An example of that has mentioned in this repo: https://github.com/mdeff/cnn_graph, which convert mnist data into a graph structure and predict the labels based on information propagation on graphs.

**Comment by shashank93jai** [comment]
@zhangyu2ustc Did you figure out how to solve this? I have a similar requirement where I have one underlying graph representation and each input is a node-wise feature representation. I'm not trying to process multiple such inputs together as a batch, So I want to be able to process inputs of size (batch_size, number of nodes, feature_dim) using a GraphConv layer but it seems that a GraphConv layer expects inputs of size (number of nodes, feature_dim). 

Is there any way to go about this? @jermainewang

**Comment by learning-chip** [comment]
> So I want to be able to process inputs of size (batch_size, number of nodes, feature_dim) using a GraphConv layer but it seems that a GraphConv layer expects inputs of size (number of nodes, feature_dim).

@shashank93jai  I got exactly the same problem. With PyTorch backend, a potential workaround is to use the JAX-like `vmap` in [functorch](https://github.com/facebookresearch/functorch) to "broadcast" over new dimensions. However, DGL is not PyTorch all the way down, but has its own sparse operator implementation (like `GSpMM` in [dgl.ops](https://docs.dgl.ai/en/0.6.x/api/python/dgl.ops.html)), which should break the PyTorch-specific `vmap`.

Also, not sure about MXNet backend (apache/incubator-mxnet##20561)

**Comment by mklpr** [comment]
> @zhangyu2ustc Did you figure out how to solve this? I have a similar requirement where I have one underlying graph representation and each input to the network is a node-wise feature representation. I'm trying to process multiple such inputs together as a batch, So I want to be able to process inputs of size (batch_size, number of nodes, feature_dim) using a GraphConv layer but it seems that a GraphConv layer expects inputs of size (number of nodes, feature_dim).
> 
> Is there any way to go about this? @jermainewang

also have this requirement, to feed a batch of samples (with different nodes/edges features and corresponding labels inputs) into a fixed graph in dgl, it seems we can construct samples like (number of nodes, batch_size, feature_dim),  and manually ensure the batch_size's dimension satisfy the normal batch behaviour in message and reduce function, below shows the demo of `batch_size = (2, 3)`.

```python
import dgl
import torch as th
import dgl.function as fn

u, v = th.tensor([0, 0, 0, 1]), th.tensor([1, 2, 3, 3])
g = dgl.graph((u, v))
g.ndata['x'] = th.ones(g.num_nodes(), 2, 3, 4)

def udf_message(edges):
    m = edges.src['x'] @ th.rand(4, 5)
    print('-----', edges.src['x'].shape, m.shape)
    return {'m' : m}

def udf_reduce(nodes):
    print('=====', nodes.mailbox['m'].shape)
    return {'y': nodes.mailbox['m'].sum(1)}

g.update_all(udf_message, udf_reduce)
print('*****', g.ndata['y'].shape)

g.update_all(fn.copy_u('x', 'm2'), fn.sum('m2', 'y2'))
print('#####', g.ndata['y2'].shape)
```

```
----- torch.Size([4, 2, 3, 4]) torch.Size([4, 2, 3, 5])
===== torch.Size([2, 1, 2, 3, 5])
===== torch.Size([1, 2, 2, 3, 5])
***** torch.Size([4, 2, 3, 5])
##### torch.Size([4, 2, 3, 4])
```

see also https://github.com/dmlc/dgl/pull/2912.
